homeassistant/util/async_.py

```diff
@@ -59,18 +60,55 @@ def run_callback() -> None:
 
 
 def check_loop() -> None:
-    """Raise if called inside the event loop."""
+    """Warn if called inside the event loop."""
     try:
         get_running_loop()
         in_loop = True
     except RuntimeError:
         in_loop = False
 
-    if in_loop:
-        _LOGGER.warning(
+    if not in_loop:
+        return
+
+    found_frame = None
+
+    for frame in reversed(traceback.extract_stack()):
+        for path in ("custom_components/", "homeassistant/components/"):
+            try:
+                index = frame.filename.index(path)
+                found_frame = frame
+                break
+            except ValueError:
+                continue
+
+        if found_frame is not None:
+            break
+
+    # Did not source from integration? Hard error.
+    if found_frame is None:
+        raise RuntimeError(
             "Detected I/O inside the event loop. This is causing stability issues. Please report issue"
         )
 
+    start = index + len(path)
+    end = found_frame.filename.index("/", start)
```

Comment:
Would it be easier to use the pathlib api to search and match parts of the filename?

Reply:
I could use relative_to but I don't think that we end up with less code.